src/checksum.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// Stateful checksum calculator.
+///
+/// Ext4 has [metadata checksums][0] for most data structures
+/// (superblock, block group descriptor, inode, etc).
+///
+/// [0]: https://www.kernel.org/doc/html/latest/filesystems/ext4/overview.html#checksums
+///
+/// # Seed
+///
+/// The default seed is `0xffff_ffff`. This is used for the superblock
+/// checksum. Other structures use a checksum seed stored in the
+/// superblock. The checksum seed is derived from the filesystem's
+/// initial UUID.
+#[derive(Clone)]
+pub(crate) struct Checksum {
+    digest: crc::Digest<'static, u32>,
+}
+
+impl Checksum {
+    /// Create a `Checksum` with the default seed.
+    pub(crate) fn new() -> Self {
+        Self::with_seed(0xffff_ffff)
+    }
+
+    /// Create a `Checksum` with the given `seed`.
+    pub(crate) fn with_seed(seed: u32) -> Self {
+        const CRC32C: crc::Crc<u32> = crc::Crc::<u32>::new(&crc::CRC_32_ISCSI);
+
+        Self {
+            digest: CRC32C.digest_with_initial(seed.reverse_bits()),
+        }
+    }
+
+    /// Extend the digest with arbitrary data.
+    pub(crate) fn update(&mut self, data: &[u8]) {
+        self.digest.update(data);
+    }
+
+    /// Extend the digest with a little-endian `u16`.
+    pub(crate) fn update_u16_le(&mut self, data: u16) {
+        self.update(&data.to_le_bytes());
+    }
+
+    /// Extend the digest with a little-endian `u32`.
+    pub(crate) fn update_u32_le(&mut self, data: u32) {
+        self.update(&data.to_le_bytes());
+    }
+
+    /// Get the final value of the checksum.
+    ///
+    /// This consumes the `Checksum`.
+    pub(crate) fn finalize(self) -> u32 {
+        self.digest.finalize() ^ (!0)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_checksum() {
+        let mut c = Checksum::new();
+        c.update_u32_le(1);
+        c.update_u32_le(2);
+        assert_eq!(c.finalize(), 0x858c13d3);
+
+        let mut c = Checksum::with_seed(123);
+        c.update_u32_le(1);
+        c.update_u32_le(2);
+        assert_eq!(c.finalize(), 0xfc527a0a);
```

Comment:
How do we know these constants are right?

Reply:
Experimentally. I got the checksums working with actual filesystem data, then added these tests with my own data and used whatever the checksum struct produced.